### PR TITLE
better argument validation

### PR DESCRIPTION
**Pull request by SimonCropp**

### CODE REVIEW

**Comment by SimonCropp** [comment]
@Particular/nservicebus review please

**Comment by SzymonPobiega** [comment]
I like the test but couldn't we push it a step forward and have a Fody plugin that emits these Guard clauses for each argument of each public (available to users) method? This will force us to follow the good practice of not ever allowing nulls as argument values.

**Comment by danielmarbach** [comment]
I think these are good changes. Validation of public API is for me a must have.

Out of curiosity. What happened to the Code Contracts library once published by MS?

Another question: Should we also use Resharper annotations inside the guard methods? Something like here https://github.com/hmemcpy/ReSharper-Secrets/blob/master/OmniLib/Guard.cs. This would then instruct resharper to not add warnings for possible NullReference etc. Haven't used them a lot and also not sure about the value add. Just wanted to mention it for completeness

**Comment by SimonCropp** [comment]
@danielmarbach 

Re Code Contracts

I have tried code contracts on several occasion and it has always cause me pain. Using it on a build server, managing the installation/updates (since it is a vs extension), the fact that it makes some bad assumptions about where to plug into the build pipeline. I even had one instance of a corrupted install that mean i could not build a project and needed to reinstall VS. If someone else feels all these are trivial and has the experience to back it up and solve the problems I am happy to defer to that experience.

> Should we also use Resharper annotations inside the guard methods

Yeah I figured this might be a logical next step.

@SzymonPobiega 

> I like the test but couldn't we push it a step forward and have a Fody plugin that emits these Guard clauses for each argument of each public (available to users) method

Building a Fody adding for proper argument validation is a vary deep problem to solve. There is https://github.com/Fody/NullGuard but it only does nulls and would also need significant testing once added to a project of the size of NSB. since we are having a more constrained public API moving forward I was happy to just do it with code and not add a new dependency on NSB

> I like the test

Also it is not a complete test ATM in  that it doesn fail for missing validation. it is more to highlight where we are missing validation. Once we get up to a point that we have good validation on the public API I think it can be handled with code reviews moving forward and the test can possibly be deleted.

@SzymonPobiega @danielmarbach 

But do you guys actually have an feedback on the validation I added rather than the approach I took? i have not changed the existing approach to validation. we already had a guard class and we already did some null checking. I just made it consistent as part of this PR. I did not make changes to tolling as part of this PR. So if you want to have extended discussion on tooling for validation perhaps raise a new issue?

**Comment by johnsimons** [comment]
The `Guard.AgainstDefault` is hard to understand IMO, I would prefer a more specific api eg, `Guard.AgainstNull`, doing a code review is hard to figure our if the arg is a reference type or not.
Anyway just something that I'm finding hard to get used to.

File: src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs [hunk 1]
```diff
@@ -28,6 +28,7 @@ internal TransactionSettings(ReadOnlySettings settings)
         /// <param name="doNotWrapHandlersExecutionInATransactionScope">Should handlers be wrapped</param>
         public TransactionSettings(bool isTransactional, TimeSpan transactionTimeout, IsolationLevel isolationLevel, bool suppressDistributedTransactions, bool doNotWrapHandlersExecutionInATransactionScope)
         {
+            Guard.AgainstNegative(transactionTimeout, "transactionTimeout");
```
Comment (johnsimons) [inline src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs]: I wonder if we should also validate the timeout passed against the machine setting ?

File: src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs [hunk 2]
```diff
@@ -32,6 +32,7 @@ internal LogicalMessage(MessageMetadata metadata, object message, Dictionary<str
         /// <param name="newInstance">The new instance.</param>
         public void UpdateMessageInstance(object newInstance)
         {
+            Guard.AgainstDefault(newInstance, "newInstance");
```
Comment (johnsimons) [inline src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs]: I wonder if it makes sense the null the body ?

File: src/NServiceBus.Core/Transports/IncomingContextExtensions.cs [hunk 3]
```diff
@@ -15,6 +15,7 @@ public static class IncomingContextExtensions
         /// <param name="address">The native transport address that should be used to address this receiver</param>
         public static void SetPublicReceiveAddress(this IncomingContext context, string address)
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/IncomingContextExtensions.cs]: address missing validation ?

File: src/NServiceBus.Core/Transports/IncomingMessage.cs [hunk 4]
```diff
@@ -31,6 +31,7 @@ public class IncomingMessage
         /// <param name="bodyStream">The message body stream</param>
         public IncomingMessage(string messageId,Dictionary<string,string> headers,Stream bodyStream)
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/IncomingMessage.cs]: I don't think stream and headers can be null

File: src/NServiceBus.Core/Timeout/ConfigurationBuilderExtensions.cs [hunk 5]
```diff
@@ -12,10 +12,11 @@ public static class ConfigurationTimeoutExtensions
         /// By default we wait for 2 seconds for the storage to come back.
         /// This method allows to change the default and extend the wait time.
         /// </summary>
-        /// <param name="config"></param>
+        /// <param name="config">The instance of <see cref="BusConfiguration"/> to extend.</param>
         /// <param name="timeToWait">Time to wait before raising a critical error.</param>
         public static void TimeToWaitBeforeTriggeringCriticalErrorOnTimeoutOutages(this BusConfiguration config, TimeSpan timeToWait)
```
Comment (johnsimons) [inline src/NServiceBus.Core/Timeout/ConfigurationBuilderExtensions.cs]: timeToWait needs to be positive or zero

File: src/NServiceBus.Core/TimeToBeReceivedAttribute.cs [hunk 6]
```diff
@@ -15,6 +15,7 @@ public sealed class TimeToBeReceivedAttribute : Attribute
         /// <param name="timeSpan">A timeSpan that can be interpreted by <see cref="TimeSpan.Parse(string)"/>.</param>
         public TimeToBeReceivedAttribute(string timeSpan)
         {
+            Guard.AgainstDefaultOrEmpty(timeSpan, "timeSpan");
```
Comment (johnsimons) [inline src/NServiceBus.Core/TimeToBeReceivedAttribute.cs]: needs to be positive

File: src/NServiceBus.Core/Settings/TransactionSettings.cs [hunk 7]
```diff
@@ -109,6 +109,7 @@ public TransactionSettings DefaultTimeout(TimeSpan defaultTimeout)
                     "Timeout requested is longer than the maximum value for this machine. Please override using the maxTimeout setting of the system.transactions section in machine.config");
             }
 
+            Guard.AgainstNegative(defaultTimeout, "defaultTimeout");
```
Comment (johnsimons) [inline src/NServiceBus.Core/Settings/TransactionSettings.cs]: move this check before the more expensive one ?

File: src/NServiceBus.Core/Serializers/XML/Config/XmlSerializationExtentions.cs [hunk 8]
```diff
@@ -57,10 +57,8 @@ public static SerializationExtentions<XmlSerializer> DontWrapRawXml(this Seriali
         /// </param>
         public static SerializationExtentions<XmlSerializer> Namespace(this SerializationExtentions<XmlSerializer> config, string namespaceToUse)
         {
-            if (string.IsNullOrEmpty(namespaceToUse))
-            {
-                throw new ConfigurationErrorsException("Can't use a null or empty string as the xml namespace");
-            }
+            Guard.AgainstDefault(config, "config");
+            Guard.AgainstDefaultOrEmpty(namespaceToUse, "namespaceToUse");
```
Comment (johnsimons) [inline src/NServiceBus.Core/Serializers/XML/Config/XmlSerializationExtentions.cs]: not sure if it is valid to have namespace as empty ?

File: src/NServiceBus.Core/Scheduling/Schedule.cs [hunk 9]
```diff
@@ -30,6 +31,7 @@ public Schedule(IBuilder builder)
         /// <param name="task">The <see cref="System.Action"/> to execute.</param>
         public void Every(TimeSpan timeSpan, Action task)
```
Comment (johnsimons) [inline src/NServiceBus.Core/Scheduling/Schedule.cs]: timespan positive

File: src/NServiceBus.Core/Scheduling/Schedule.cs [hunk 10]
```diff
@@ -49,6 +51,8 @@ public void Every(TimeSpan timeSpan, Action task)
         /// <param name="name">The name to use used for logging inside the new <see cref="Thread"/>.</param>
         public void Every(TimeSpan timeSpan, string name, Action task)
```
Comment (johnsimons) [inline src/NServiceBus.Core/Scheduling/Schedule.cs]: timespan positive

File: src/NServiceBus.Core/Serializers/XML/Config/XmlSerializationExtentions.cs (hunk 8 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Serializers/XML/Config/XmlSerializationExtentions.cs]: yep null is valid. fixed

File: src/NServiceBus.Core/Scheduling/Schedule.cs (hunk 9 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Scheduling/Schedule.cs]: fixed

File: src/NServiceBus.Core/Scheduling/Schedule.cs (hunk 10 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Scheduling/Schedule.cs]: fixed

File: src/NServiceBus.Core/Settings/TransactionSettings.cs (hunk 7 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Settings/TransactionSettings.cs]: fixed

File: src/NServiceBus.Core/TimeToBeReceivedAttribute.cs (hunk 6 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/TimeToBeReceivedAttribute.cs]: fixed

File: src/NServiceBus.Core/Timeout/ConfigurationBuilderExtensions.cs (hunk 5 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Timeout/ConfigurationBuilderExtensions.cs]: fixed

File: src/NServiceBus.Core/Transports/IncomingContextExtensions.cs (hunk 3 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/IncomingContextExtensions.cs]: fixed

File: src/NServiceBus.Core/Transports/IncomingMessage.cs (hunk 4 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/IncomingMessage.cs]: fixed

File: src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs]: doesnt null only make sense for a control message?

File: src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs (hunk 1 above)
Comment (SimonCropp) [inline src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs]: meaning?

**Comment by SimonCropp** [comment]
@johnsimons 

> The Guard.AgainstDefault is hard to understand IMO,

fixed

File: src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs (hunk 2 above)
Comment (johnsimons) [inline src/NServiceBus.Core/Unicast/Messages/LogicalMessage.cs]: I think that depends on the user, they control it, not sure if our code
would work though

File: src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs (hunk 1 above)
Comment (johnsimons) [inline src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs]: Here https://github.com/Particular/NServiceBus/blob/36e9c56d0f2e4f251e9f4d76eb22bd85fafd8b50/src/NServiceBus.Core/Settings/TransactionSettings.cs#L104-L135

Comment (SimonCropp) [inline src/NServiceBus.Core/Unicast/Transport/TransactionSettings.cs]: i dont know? should we?

**Comment by SeanFeldman** [comment]
@SimonCropp would it make sense to have `Guard` class available outside of core for downstream projects as well?

**Comment by SimonCropp** [comment]
@SeanFeldman nope. Not worth leaking the public api

**Pull request merged**